### DISCUSSION TITLE
Documentation Wrong/Misleading Regarding Sendable Conformance

### DISCUSSION THREAD

**Discussion by philipbel** [pointfreeco/swift-dependencies#245] [General]
### Description

Reading on TCA/Dependencies, I stumbled upon this requirement for dependencies to be `Sendable`. Am I missing something?

`DependencyKey` (via `TestDependencyKey`) requires that `Value` conform to `Sendable`.

Reading the [Protocol-based dependencies](https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/designingdependencies#Protocol-based-dependencies) section of the documentation, this example is wrong (or misleading at the very least):

```Swift
struct LiveAudioPlayer: AudioPlayer { // 🛑 Type 'any AudioPlayer' does not conform to the 'Sendable' protocol
  let audioEngine: AVAudioEngine // 🛑 Static property 'liveValue' is not concurrency-safe because it is not either conforming to 'Sendable' or isolated to a global actor; this is an error in Swift 6
  // ...
}
// ...
private enum AudioPlayerKey: DependencyKey {
  static let liveValue: any AudioPlayer = LiveAudioPlayer()
  // ...
}
```

`LiveAudioPlayer` is not `Sendable` and can't be easily made `Sendable`, as `AVAudioEngine` is a class.

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

Documentation should clearly state the `Sendable` requirements and provide guidelines how to isolate to a global actor, if that makes sense (too new to TCA, not even sure whether isolating to `MainActor` is possible).

### Actual behavior

Dependencies must be `Sendable` or the compiler generates a warning (presumably an error in Swift 6).

### Steps to reproduce

1. Copy-paste the provided examples
2. Generates a warning

### Dependencies version information

1.3.2

### Destination operating system

iOS 17

### Xcode version information

15.4

### Swift Compiler version information

```shell
▶ xcrun swiftc --version
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: x86_64-apple-macosx14.0
```

**Comment by mbrandonw** [comment]
Hi @philipbel, it is true that dependencies are required to be sendable because ultimately they are stored in a `@TaskLocal`. However, a lot of the docs were written before concurrency warnings got stricter and before the dust had started to settle on which APIs in Apple's ecosystem would be concurrency safe and which would not.

If `AVAudioEngine` is truly never going to send sendable then we will need to change the example a bit. The modification is quite easy, you will just wrap the `AVAudioEngine` in an actor or mutex/lock to protect it, and then access it within the implementation of the dependency.

Since Swift 6 is still not officially out, and who knows how sendability is going to settle in Apple frameworks, I am going to convert this to a discussion for now. As we get closer to the final release of Swift 6 we will update the docs accordingly and make sure it is clear that dependencies must be sendable.

**Reply by daksh-dls** [reply]
It is out now and please update the docs @mbrandonw 
Thanks!

**Reply by PereAlmendro** [reply]
Hi @mbrandonw @stephencelis

Is there any official solution about dependencies conforming Sendable for swift 6 ? 

Thanks in advance.

**Reply by mbrandonw** [reply]
Dependencies must always conform to `Sendable` and it has always been that way and will never change. The `@TaskLocal` forces that and there is no way around it.

The only thing we could do now is update the `AudioEngine` example in the docs, but regardless… dependencies must _always_ be sendable.
